rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,633 @@
+for renewals of the same certificate lineage.
+
+Running bot instances will establish a filesystem lock on the bot data
+directory to ensure only a single local instance can be started at any time.
+Additional renewal attempts using the same credentials with a mismatched
+certificate generation counter are presumed to be malicious and will
+immediately lock the bot user, preventing further renewal.
+
+Additionally, to further restrict renewals, impersonated certificates will be
+generated with a "disallow-reissue" flag to entirely prevent malicious
+renewals of a compromised secondary / end-user certificate.
+
+Lastly, regardless of the initial certificate's TTL, we can have the auth
+server ensure that, when renewing, renewed certificate TTLs may only decrease
+in length. This would prevent an attacker from stealing a renewable certificate
+and immediately requesting a new certificate with the auth server's max
+renewable TTL.
+
+#### Artifacts
+
+For internal use, the bot generates a set of renewable TLS certificates.
+This set of credentials is only used certificate renewal, user certificate
+generation, and establishing CA rotation watches. These artifacts should be
+persisted to some private location to allow the bot to recover after a process
+restart.
+
+| Kind                   | Path         | Notes                            |
+|------------------------|--------------|----------------------------------|
+| TLS CA certificates    | `tlscacerts` | Teleport cluster CA certs (both) |
+| TLS client certificate | `tlscert`    | Signed by Teleport `UserCA`      |
+| Private key (TLS)      | `key`        |                                  |
+
+The bot then generates a set of non-renewable credentials for each configured
+credential destination. In the initial implementation, the following artifacts
+may or may not be generated depending on the user's configuration (kind and
+templates).
+
+| Kind                   | Path          | Notes                                   | Kind | Config Template
+| -----------------------|---------------|-----------------------------------------|------|---------
+| Private key            | `key`         | Shared for both SSH and TLS uses        | Both | Always
+| SSH public key         | `key.pub`     | Required for OpenSSH compatibility      | SSH  | Always
+| SSH client certificate | `sshcert`     | Signed by Teleport `UserCA`             | SSH  | Always
+| SSH known hosts        | `known_hosts` | Teleport `HostCA` certs, OpenSSH format | SSH  | `ssh-client`
+| SSH client config      | `ssh_config`  | `Include`-ready SSH client config       | SSH  | `ssh-client`
+| TLS CA certificates    | `tlscacerts`  | Teleport CA certificates (both)         | TLS  | Always
+| TLS client certificate | `tlscert`     | Signed by Teleport `UserCA`             | TLS  | Always
+
+In the case of `directory` destinations, we encourage the use of POSIX ACLs.
+This is meant to ensure that the primary (renewable) credentials are
```

Comment:
Actively checking this when writing to them and throwing a warning/event is advisable.

Reply:
I'll add a note to warn about this. I'd like to make it a harder error but unfortunately we can't depend on ACLs being supported on any given filesystem / OS.